pkg/retainer/txretainer/database.go

```diff
@@ -0,0 +1,182 @@
+package txretainer
+
+import (
+	"gorm.io/gorm"
+
+	"github.com/iotaledger/hive.go/ierrors"
+	iotago "github.com/iotaledger/iota.go/v4"
+)
+
+var (
+	ErrEntryNotFound = ierrors.New("entry not found")
+
+	dbTables = append([]interface{}{
+		&Status{},
+	}, metadataTables...)
+
+	metadataTables = []interface{}{
+		&txMetadata{},
+	}
+)
+
+func (r *TransactionRetainer) handleError(err error, message string) error {
+	if err != nil {
+		err = ierrors.Wrap(err, message)
+		r.errorHandler(err)
+	}
+
+	return err
+}
+
+func (r *TransactionRetainer) IsInitialized() bool {
+	hasTable := false
+
+	err := r.dbExecFunc(func(db *gorm.DB) error {
+		hasTable = db.Migrator().HasTable(&Status{})
+		return nil
+	})
+	r.handleError(err, "failed to check if store is initialized")
+
+	return hasTable
+}
+
+func (r *TransactionRetainer) CreateTables() error {
+	err := r.dbExecFunc(func(db *gorm.DB) error {
+		return db.Migrator().CreateTable(dbTables...)
+	})
+
+	return r.handleError(err, "failed to create tables")
+}
+
+func (r *TransactionRetainer) AutoMigrate() error {
+	err := r.dbExecFunc(func(db *gorm.DB) error {
+		// Create the tables and indexes if needed
+		return db.AutoMigrate(dbTables...)
+	})
+
+	return r.handleError(err, "failed to auto migrate tables")
+}
+
+func (r *TransactionRetainer) Status() (*Status, error) {
+	status := &Status{}
+
+	if err := r.dbExecFunc(func(db *gorm.DB) error {
+		return db.Take(&status).Error
+	}); err != nil {
+		if ierrors.Is(err, gorm.ErrRecordNotFound) {
+			return nil, ErrEntryNotFound
+		}
+
+		return nil, r.handleError(err, "failed to get status")
+	}
+
+	return status, nil
+}
+
+func (r *TransactionRetainer) Clear() error {
+	err := r.dbExecFunc(func(db *gorm.DB) error {
+		// drop all tables
+		if err := db.Migrator().DropTable(dbTables...); err != nil {
+			return err
+		}
+
+		// re-create tables
+		return db.Migrator().CreateTable(dbTables...)
+	})
+
+	return r.handleError(err, "failed to clear database")
+}
+
+func (r *TransactionRetainer) CloseDatabase() error {
+	err := r.dbExecFunc(func(db *gorm.DB) error {
+		sqlDB, err := db.DB()
+		if err != nil {
+			return err
+		}
+
+		return sqlDB.Close()
+	})
+
+	return r.handleError(err, "failed to close database")
+}
+
+func (r *TransactionRetainer) updateTxMetadata(txID iotago.TransactionID, newTxMeta *txMetadata) {
+	err := r.dbExecFunc(func(db *gorm.DB) error {
+		return db.Transaction(func(tx *gorm.DB) error {
+			txMeta := &txMetadata{}
+
+			// init the txMeta if it does not exist yet, or get the existing one from the database
+			tx.FirstOrInit(txMeta, &txMetadata{TransactionID: txID[:]})
```

Comment:
```suggestion
			dbTx.FirstOrInit(txMeta, &txMetadata{TransactionID: txID[:]})
```

Calling this `dbTx` makes this slightly more understandable with the mix of tx terminology from the db and ours.